leopard/leopard.nim

```diff
@@ -0,0 +1,263 @@
+  ## `recovered`  - list of recovered `buffers` of size `bufSize`
+  ##
+
+  if data.len != self.buffers:
+    return err("Number of data buffers should match!")
+
+  if parity.len != self.parity:
+    return err("Number of parity buffers should match!")
+
+  if recovered.len != self.buffers:
+    return err("Number of recovered buffers should match buffers!")
+
+  # clean out work and data buffers
+  for i in 0..<self.workBufferCount:
+    zeroMem(self.workBufferPtr[i], self.bufSize)
+
+  for i in 0..<self.decodeBufferCount:
+    zeroMem(self.decodeBufferPtr[i], self.bufSize)
+
+  for i in 0..<data.len:
+    zeroMem(self.dataBufferPtr[i], self.bufSize)
+
+  # this is needed because erasures are nil pointers
+  var
+    dataPtr = newSeq[LeoBufferPtr](data.len)
+    parityPtr = newSeq[LeoBufferPtr](self.workBufferCount)
+
+  # copy data into aligned buffer
+  for i in 0..<data.len:
+    if data[i].len > 0:
+      copyMem(self.dataBufferPtr[i], addr data[i][0], self.bufSize)
+      dataPtr[i] = self.dataBufferPtr[i]
+    else:
+      dataPtr[i] = nil
+
+  # copy parity into aligned buffer
+  for i in 0..<self.workBufferCount:
+    if i < parity.len and parity[i].len > 0:
+      copyMem(self.workBufferPtr[i], addr parity[i][0], self.bufSize)
+      parityPtr[i] = self.workBufferPtr[i]
+    else:
+      parityPtr[i] = nil
+
+  let
+    res = leo_decode(
+      self.bufSize.cuint,
+      self.buffers.cuint,
+      self.parity.cuint,
+      self.decodeBufferCount.cuint,
+      cast[ptr pointer](addr dataPtr[0]),
+      cast[ptr pointer](addr parityPtr[0]),
+      cast[ptr pointer](addr self.decodeBufferPtr[0]))
+
+  if ord(res) != ord(LeopardSuccess):
+    return err(leoResultString(res.LeopardResult))
+
+  for i, p in dataPtr:
+    if p.isNil:
+      copyMem(addr recovered[i][0], self.decodeBufferPtr[i], self.bufSize)
+
+  ok()
+
+func free*(self: var Leo) =
+  if self.workBufferPtr.len > 0:
+    for i, p in self.workBufferPtr:
+      if not isNil(p):
+        p.leoFree()
+        self.workBufferPtr[i] = nil
+
+    self.workBufferPtr.setLen(0)
+
+  if self.dataBufferPtr.len > 0:
+    for i, p in self.dataBufferPtr:
+      if not isNil(p):
+        p.leoFree()
+        self.dataBufferPtr[i] = nil
+
+    self.dataBufferPtr.setLen(0)
+
+  if self.kind == LeoCoderKind.Decoder:
+    if self.decodeBufferPtr.len > 0:
+      for i, p in self.decodeBufferPtr:
+        if not isNil(p):
+          p.leoFree()
+          self.decodeBufferPtr[i] = nil
+      self.decodeBufferPtr.setLen(0)
+
+# TODO: The destructor doesn't behave as
+# I'd expect it, it's called many more times
+# than it should. This is however, most
+# likely my misinterpretation of how it should
+# work.
+# proc `=destroy`*(self: var Leo) =
+#   self.free()
+
+proc init[TT: Leo](
+  T: type TT,
+  bufSize,
+  buffers,
+  parity: int,
+  kind: LeoCoderKind): Result[T, cstring] =
+  if bufSize mod BuffMultiples != 0:
+    return err("bufSize should be multiples of 64 bytes!")
+
```

Comment:
Should we add a check (and a test in `tests/testleopard.nim`) that `buffers + parity <= 65536`?

Reply:
Good idea!